### PR TITLE
Add support for proc-macro crates, add serde_derive

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: build_extra/rust/run.py [hunk 1]
```diff
@@ -7,19 +7,32 @@
...
 import re
 
+if sys.platform == 'win32':
+    # On Windows, when gn is setting up the build toolchain, it produces a set
+    # of environment variables that are required to invoke the right build
+    # toolchain. We need to load those environment variables here too in order
+    # for rustc to be able to successfully invoke the linker tool.
+    # The file is in 'windows environment block' format, which contains
+    # multiple 'key=value' pairs, separated by '\0' bytes, and terminated by
+    # two '\0' bytes at the end.
```
Comment (ry) [inline build_extra/rust/run.py]: > separated by '\0' bytes

omg

File: build_extra/rust/rust.gni [hunk 2]
```diff
@@ -228,7 +255,7 @@ template("rust_crate") {
       foreach(c, cfg) {
         args += [
           "--cfg",
-          "c",
+          c,
```
Comment (ry) [inline build_extra/rust/rust.gni]: Was this an error?

File: cli/Cargo.toml [hunk 3]
```diff
@@ -36,10 +36,12 @@ regex = "1.1.6"
 remove_dir_all = "0.5.1"
 ring = "0.14.6"
 rustyline = "4.0.0"
+serde = "1.0.91"
+serde_derive = "1.0.91"
 serde_json = "1.0.39"
 source-map-mappings = "0.5.0"
 tempfile = "3.0.7"
-tokio = "0.1.19"
+tokio = "0.1.20"
```
Comment (ry) [inline cli/Cargo.toml]: When you land this, be sure to separate the "crate upgrade" commit from prog-macro additions.

File: cli/main.rs [hunk 4]
```diff
@@ -52,6 +55,12 @@ use futures::Future;
 use log::{LevelFilter, Metadata, Record};
 use std::env;
 
+#[derive(Serialize, Deserialize, Debug)]
+struct Point {
+  x: i32,
+  y: i32,
+}
```
Comment (ry) [inline cli/main.rs]: Reminder to remove this before landing.

File: tools/format.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ def qrun(cmd, env=None):
      find_exts(["core"], [".cc", ".h"]))
 
 print "gn format"
-for fn in ["BUILD.gn", ".gn"] + find_exts(["build_extra", "core"],
+for fn in ["BUILD.gn", ".gn"] + find_exts(["build_extra", "cli", "core"],
```
Comment (ry) [inline tools/format.py]: Oops. Let's do this as a separate commit too.

File: build_extra/rust/rust.gni (hunk 2 above)
Comment (piscisaureus) [reply build_extra/rust/rust.gni]: yes
